### PR TITLE
nixos/your_spotify: init at 1.10.1

### PR DESCRIPTION
**Pull request by PatrickDaG**
## Description of changes
- This PR adds your_spotify a self hosted spotify dashboard.
This includes package, module and nixos tests.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by PatrickDaG** [comment]
Sadly the upstream mongo does not compile currently. I've tested it with a binary mongodb using the prebuild debian packages but this probably can't  be upstreamed and since the package depends on mongodb this means it can't be tested using current nixpkgs.

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    };
+
+    package = mkPackageOption pkgs "your_spotify" {};
+
+    clientPackage = mkOption {
+      type = package;
+      default = cfg.package.client.override {apiEndpoint = cfg.settings.API_ENDPOINT;};
+      description = "Client package to use.";
+    };
+    spotifyPublicFile = mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/your_spotify.nix:54]: If this is a public key, it could also safely be written to the nix store.

File: pkgs/by-name/yo/your_spotify/package.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      runHook preBuild
+      pushd ./apps/server/
+      yarn --offline run build
+      popd
+      runHook postBuild
+    '';
+    nativeBuildInputs = [makeWrapper yarn prefetch-yarn-deps];
+    installPhase = ''
+      mkdir -p $out
+      cp -r node_modules $out/node_modules
```
Comment (SuperSandro2000) [inline pkgs/by-name/yo/your_spotify/package.nix:51]:
```suggestion
      cp -r node_modules $out/share/your-spotify/node_modules
```

We should keep $out a bit clean

File: pkgs/by-name/yo/your_spotify/package.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  src = fetchFromGitHub {
+    owner = "Yooooomi";
+    repo = "your_spotify";
+    rev = "refs/tags/${version}";
+    hash = "sha256-umm7J5ADY2fl+tjs6Qeda5MX2P55u0eCqwW+DWLK8Kc=";
+  };
+  client = callPackage ./client.nix {inherit src version;};
+in
+  mkYarnPackage rec {
+    inherit version src;
```
Comment (SuperSandro2000) [inline pkgs/by-name/yo/your_spotify/package.nix:22]: Please use on some+version

File: pkgs/by-name/yo/your_spotify/client.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    mkdir -p $out
+    cp -r ./apps/client/build/* $out
+    substituteInPlace $out/variables-template.js --replace-quiet '__API_ENDPOINT__' "${apiEndpoint}"
+    mv $out/variables-template.js $out/variables.js
+  '';
+  doDist = false;
+  meta = with lib; {
+    homepage = "https://github.com/Yooooomi/your_spotify";
+    changelog = "https://github.com/Yooooomi/your_spotify/releases/tag/${version}";
+    description = "YourSpotify is a self-hosted application that tracks what you listen and offers you a dashboard to explore statistics about it!";
```
Comment (SuperSandro2000) [inline pkgs/by-name/yo/your_spotify/client.nix:50]: We can inherit this from the main package

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      script = ''
+        export SPOTIFY_PUBLIC=$(< "$CREDENTIALS_DIRECTORY/SPOTIFY_PUBLIC")
+        export SPOTIFY_SECRET=$(< "$CREDENTIALS_DIRECTORY/SPOTIFY_SECRET")
+        ${pkgs.your_spotify}/bin/your_spotify_migrate
+        exec ${pkgs.your_spotify}/bin/your_spotify_server
+      '';
+      serviceConfig = {
+        User = "your_spotify";
+        Group = "your_spotify";
+        DynamicUser = true;
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix:126]: It might be a bit more complicated but even without DynamicUser you would need to use some sort of indirection to prevent the secret from ending up in the nix-store. And the current way prevents you from having to write an environment file yourself and you don't hide which variables are set by the secret.

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [
+          "@system-service"
+          "@pkey"
+        ];
+        UMask = "0077";
+      };
+      wantedBy = ["multi-user.target"];
+    };
+    services.nginx = mkIf (cfg.nginxVirtualHost != null) {
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix:172]: It works as is and I didn't want to specify more than necessary to not make assumption about end users network setup.

File: pkgs/by-name/yo/your_spotify/client.nix [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  makeWrapper,
+  mkYarnPackage,
+  prefetch-yarn-deps,
+  src,
+  version,
+  yarn,
+}:
+mkYarnPackage rec {
+  inherit version src;
+  name = "your_spotify_client-${version}";
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/client.nix:14]: The problem it that `mkYarnPackage` takes the name from the package.json located in the source folder. Due to the repository layout using npms workspaces this name is `your_spotify/root` which results in the package name generated being `your_spotify-root-1.8.0`. I wanted to avoid the root infix and thus set the name directly. But if the infix is less undesirable than using `name` directly I can change it of course.

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        PrivateDevices = true;
+        StateDirectoryMode = "0700";
+        Restart = "always";
+
+        LoadCredential = ["SPOTIFY_PUBLIC:${cfg.spotifyPublicFile}" "SPOTIFY_SECRET:${cfg.spotifySecretFile}"];
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        LockPersonality = true;
+        #MemoryDenyWriteExecute = true;
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix:140]: I added a comment that this is disabled due to V8 JIT.

File: pkgs/by-name/yo/your_spotify/package.nix (hunk 3 above)
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/package.nix:22]: What do you mean by some?

File: nixos/modules/services/web-apps/your_spotify.nix (hunk 1 above)
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix]: Done

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.your_spotify = {
+      after = ["network.target"];
+      script = ''
+        export SPOTIFY_PUBLIC=$(< "$CREDENTIALS_DIRECTORY/SPOTIFY_PUBLIC")
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix]: Done

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        export SPOTIFY_SECRET=$(< "$CREDENTIALS_DIRECTORY/SPOTIFY_SECRET")
+        ${pkgs.your_spotify}/bin/your_spotify_migrate
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix]: Done

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        StateDirectoryMode = "0700";
+        Restart = "always";
+
+        LoadCredential = ["SPOTIFY_PUBLIC:${cfg.spotifyPublicFile}" "SPOTIFY_SECRET:${cfg.spotifySecretFile}"];
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        LockPersonality = true;
+        #MemoryDenyWriteExecute = true;
+        #NoNewPrivileges = true; # Implied by DynamicUser
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix]: Done

File: nixos/modules/services/web-apps/your_spotify.nix [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        PrivateUsers = true;
+        #PrivateTmp = true; # Implied by DynamicUser
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = false; # breaks bwrap
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply nixos/modules/services/web-apps/your_spotify.nix]: These were incorrect. I've updated them.

File: pkgs/by-name/yo/your_spotify/client.nix [hunk 13]
```diff
@@ -0,0 +1,54 @@
+{
+  apiEndpoint ? "localhost:8080",
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/client.nix]: It's listening on both so this is fine.

File: pkgs/by-name/yo/your_spotify/package.nix [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        --add-flags "$out/lib/index.js"
+    '';
+    doDist = false;
+    passthru = {
+      inherit client;
+    };
+    meta = with lib; {
+      homepage = "https://github.com/Yooooomi/your_spotify";
+      changelog = "https://github.com/Yooooomi/your_spotify/releases/tag/${version}";
+      description = "YourSpotify is a self-hosted application that tracks what you listen and offers you a dashboard to explore statistics about it!";
```
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/package.nix]: Done.

File: pkgs/by-name/yo/your_spotify/package.nix (hunk 2 above)
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/package.nix]: Done

File: pkgs/by-name/yo/your_spotify/client.nix (hunk 4 above)
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/client.nix]: Done

**Comment by PatrickDaG** [comment]
I've addressed all comments and fixed all build errors.

File: pkgs/by-name/yo/your_spotify/package.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/by-name/yo/your_spotify/package.nix:22]: *pname

File: pkgs/by-name/yo/your_spotify/client.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/by-name/yo/your_spotify/client.nix:14]: Then we can set name manually but we should definitely set pname+version

File: pkgs/by-name/yo/your_spotify/client.nix [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    runHook postConfigure
+  '';
+  buildPhase = ''
+    runHook preBuild
+    pushd ./apps/client/
+    yarn --offline run build
+    export API_ENDPOINT="${apiEndpoint}"
+    substituteInPlace scripts/run/variables.sh --replace-quiet '/app/apps/client/' "./"
+    patchShebangs --build ./scripts/run/variables.sh
+    cat ./scripts/run/variables.sh
```
Comment (SuperSandro2000) [inline pkgs/by-name/yo/your_spotify/client.nix:37]:
```suggestion
```

File: pkgs/by-name/yo/your_spotify/client.nix [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    chmod +x ./scripts/run/variables.sh
```
Comment (SuperSandro2000) [inline pkgs/by-name/yo/your_spotify/client.nix:38]: Don't we need to first chmod it for patchshebangs to work?

Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/client.nix:37]: True. Fixed it.

File: pkgs/by-name/yo/your_spotify/package.nix (hunk 3 above)
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/package.nix:22]: I've set pname on both packages.

File: pkgs/by-name/yo/your_spotify/client.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/client.nix:14]: I've set pname and version on both packages now.

**Comment by PatrickDaG** [comment]
Updated to newest version

File: pkgs/by-name/yo/your_spotify/package.nix [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    '';
+    doDist = false;
+    passthru = {
+      inherit client;
+    };
+    meta = with lib; {
+      homepage = "https://github.com/Yooooomi/your_spotify";
+      changelog = "https://github.com/Yooooomi/your_spotify/releases/tag/${version}";
+      description = "Self-hosted application that tracks what you listen and offers you a dashboard to explore statistics about it";
+      license = licenses.gpl3;
```
Comment (Lurkki14) [inline pkgs/by-name/yo/your_spotify/package.nix:75]: `licenses.gpl3` is deprecated, use `gpl3Plus` or `gpl3Only`

**Review by Lurkki14 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/yo/your_spotify/package.nix (hunk 17 above)
Comment (PatrickDaG) [reply pkgs/by-name/yo/your_spotify/package.nix]: Changed it.

File: pkgs/by-name/yo/your_spotify/package.nix [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+      rm -r node_modules
+      export NODE_ENV="production"
+      yarn install --offline --frozen-lockfile --ignore-platform --ignore-scripts --no-progress --non-interactive
+      patchShebangs node_modules/
+
+      runHook postBuild
+    '';
+    nativeBuildInputs = [makeWrapper yarn prefetch-yarn-deps];
+    installPhase = ''
```
Comment (drupol) [inline pkgs/by-name/yo/your_spotify/package.nix:57]: If you are overriding `configurePhase`, `buildPhase`, `checkPhase`, `installPhase` or any other phase, you should not forget about explicitly running pre and post-phase hooks like [their](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L953) [original](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1008) [definitions](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1037) [do](https://github.com/NixOS/nixpkgs/blob/d71a03ad695407dd482ead32d3eddff50092a1c3/pkgs/stdenv/generic/setup.sh#L1078).

It is generally expected that an appropriate pre-phase hook (e.g. `preBuild`) hook will run at the beginning of a phase (e.g. `buildPhase`) and post-phase hook (e.g. `postBuild`) will run at the end. Hooks are normally ran as a part of a phase so if you override a phase as a whole, you will need to add `runHook hookName` calls (e.g. `runHook preBuild`) manually.

Having phases run pre/post-phase hooks is important because many setup hooks insert their own code into them – omitting a hook might therefore prevent some setup hooks required for proper functionality of a package from running. Additionally, hooks are often inserted by developers into the package expression and by users when overriding a package using `overrideAttrs`. Not running them can thus cause confusion why their code is not executed.

## Examples

### Before

```nix
  installPhase = ''
    your commands
  '';
```

### After

```nix
  installPhase = ''
    runHook preInstall

    your commands

    runHook postInstall
  '';
```

Source: https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/missing-phase-hooks.md

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/tests/all-tests.nix [hunk 19]
```diff
@@ -993,6 +993,7 @@ in {
   xterm = handleTest ./xterm.nix {};
   xxh = handleTest ./xxh.nix {};
   yabar = handleTest ./yabar.nix {};
+  your_spotify = handleTest ./your_spotify.nix {};
```
Comment (AndersonTorres) [inline nixos/tests/all-tests.nix:996]: nixos tests should be in its own commit.

Comment (PatrickDaG) [reply nixos/tests/all-tests.nix:996]: Should be fixed now.

**Review by drupol — CHANGES_REQUESTED** [review]
Eval broken

**Comment by PatrickDaG** [comment]
Fixed build error. I needed to copy the `package.json` into nixpkgs as Hydra has a problem with IFD which I didn't know cause on my machine it worked perfectly. Should be finished now.

**Review by drupol — APPROVED** [review]
